### PR TITLE
Fix infinite climb bug when using EK3_OGN_HGT_MASK

### PR DESCRIPTION
**Pull request by peterbarker**
This bug-finding and fix are not mine.  The autotest is mine, however :-)

@pavloblindnology reported them here: https://github.com/ArduPilot/ardupilot/issues/24039

@pavloblindnology was this the fix you were recommending?

This PR is based on https://github.com/ArduPilot/ardupilot/pull/26916 which adds some infrastructure changes to make autotesting with baro drift easier.

### CODE REVIEW

**Comment by rmackay9** [comment]
@peterbarker,

I guess this is a bug that can happen on a real vehicle as well?  It's not just a SITL issue is it?

**Comment by peterbarker** [comment]
Some pictures from a log where we never install the `check_altitude` message hook (pre-fix):
image

**Comment by pavloblindnology** [comment]
@peterbarker Sorry for late response. Yes, the fix is right.

> I guess this is a bug that can happen on a real vehicle as well? It's not just a SITL issue is it?

@rmackay9 I tested this in SITL only, but I don't see any reason why it can't happen on a real vehicle. I don't remember all the details of my tests as it was quite a while ago. But the main idea is that with `EK3_OGN_HGT_MASK=5`, Ardupilot doesn't use previously computed value of `ekfGpsRefHgt` to convert new GPS measurements. And as barometer altitude usually rises faster than GPS altitude, this results in constant `ekfGpsRefHgt` monotonical correction.

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 1]
```diff
@@ -684,11 +684,7 @@ void NavEKF3_core::readGpsData()
     if (validOrigin) {
         gpsDataNew.lat = gpsloc.lat;
         gpsDataNew.lng = gpsloc.lng;
-        if ((frontend->_originHgtMode & (1<<2)) == 0) {
-            gpsDataNew.hgt = (ftype)((double)0.01 * (double)gpsloc.alt - ekfGpsRefHgt);
-        } else {
-            gpsDataNew.hgt = 0.01 * (gpsloc.alt - EKF_origin.alt);
-        }
+        gpsDataNew.hgt = (ftype)((double)0.01 * (double)gpsloc.alt - ekfGpsRefHgt);
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:687]: this removes the (1<<2) bit check on the input side, do we need to change the two on the output side?

**Comment by davidbuzz** [comment]
@priseborough

**Comment by peterbarker** [comment]
Note that my autotest only reproduces the issue - I didn't point any fail condition in for the specific bug

**Review by priseborough — COMMENTED** [review]
I've forked this PR branch here https://github.com/priseborough/ardupilot/commits/PR26917/ and pushed an alternative set of patches for this issue that preserve the intent of the EK3_OGN_HGT_MASK bit 2 option.

priseborough@a38c01a
priseborough@f6a56d4

It passes the new autotests on my PC.

**Comment by peterbarker** [comment]
I've rebased this PR on master and removed the commit/revert pair.

So some less-than-fortunate news.  Somehow tridge's PR to use a better atmospheric model (https://github.com/ArduPilot/ardupilot/pull/26915) *also* fixed the autotest.

I bisected the code to work that out:
image

IOW the autotest passes before the fixes in this branch are applied, which is Less Than Good.  OTOH, since I've no idea how tridge's patch influences things, I'm quite happy to just carry on.

@priseborough are you OK with merging this even 'though the problem doesn't reproduce on master post-tridge-patches?

**Comment by priseborough** [comment]
Will check log and get back to you

**Review by priseborough — APPROVED** [review]

**Comment by peterbarker** [comment]
> The better atmospheric model changes how the offset between baro height and GPs height varies so it makes sense the autotest would be sensitive to this. I'm happy to merge on the basis the old model shows it failing before. It should be possible to make it fail on the new model by fudging the baro scale factor or similar.

Thanks Paul.  Marking as MergeOnCIPass.

**Comment by peterbarker** [comment]
Thanks @pavloblindnology - we've merged a fix for the bug you found.

**Comment by ddomit** [comment]
Any idea to which version of arducopter is going to include this?

**Comment by rmackay9** [comment]
Hi @ddomit,

This PR's fix will most likely be included in 4.5.5.

**Comment by ddomit** [comment]
> Hi @ddomit,
> 
> This PR's fix will most likely be included in 4.5.5. We would normally have included it in the next release (4.5.4) but that's likely to include only a single small fix

Thanks!! Do you  think this fix and update will solve the issue im having? Will it update the home location with the altitude corrections it does?

**Comment by rmackay9** [comment]
@ddomit,

I think it's best to move this to [the support forums](https://discuss.ardupilot.org/c/arducopter/copter-45/174).  Support within a PR is difficult and the issue may very well be unrelated.  Could you create a new topic in the support forum?

**Comment by rmackay9** [comment]
Hi @ddomit,

If bit2 ("Apply corrections to local position") is set then the vehicle's current altitude estimate is adjusted (instead of adjusting the EKF origin's altitude).  The EKF doesn't have access to the AHRS home location (including its altitude) so it is not directly adjusting it.  To put it another way, if bit2 is set and the EKF adjusts the vehicle's current altitude, before and after a flight, if you asked the AHRS what the home lat, lon and alt are it should show that they have not changed.

**Comment by postelzhuk** [comment]
Hi everyone. I made some SITL-tests from master, Plane-4.4, Plane-4.5. Almost sure this changes also fixes problem with baro drift: 
https://discuss.ardupilot.org/t/ekf3-use-barometer-and-gps-for-altitude-posz/119540/2
`EK3_OGN_HGT_MASK, 5` does help with SLOW baro drifts after this changes.

**Pull request merged**